### PR TITLE
File store basis

### LINKED ISSUE #13632: File map storage: Basic storage
## Description

Create a file storage implementation which covers (almost) all areas.

Features included:
- YAML format of files
- Files are stored in a structure within a given directory

Features excluded:
- No-downtime upgrade support
- Concurrent access
- Authorization
- Single-use tokens

### LINKED ISSUE #15885: Add write ability to file store

### PR DESCRIPTION
**Pull request by hmlnarik**
This PR introduces file store basis. It includes generic file store based on YAML in a directory structure, and Keycloak is able to run with file store for almost all areas, see [Q&A](https://github.com/keycloak/keycloak/discussions/16768). The limitations are:

- Single-use objects store is not supported due to different nature (`SingleUseObjectProvider` is a global store, has no realm parameter, and thus does not fit into `/<realm-id>/<area>/object.yaml` structure).
- There is only a limited support for authz (will be tackled in #13639) 
- Path traversal is still possible and will be limited in a separate issue: #17029
- `clientRole` flag is wrongly treated as a field and leads to warnings in log. This will be treated in #13631 or perhaps sooner (#16857).
- Has no support for locking (will be treated in #15837)
- Has no support for no-downtime upgrade (will be tackled in individual issues, see #13624 for their list)

Fixes: #13632
Fixes: #15885

### CODE REVIEW

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java [hunk 1]
```diff
@@ -531,25 +531,25 @@ public void generate(TypeElement e) throws IOException {
...
-                          return;
-                      }
-                      TypeMirror fieldType = determineFieldType(originalField, methodsPerAttribute.get(originalField));
-                      String field = fieldsClassName + "." + toEnumConstant(originalField);
+                    String originalField = m2field.get(ee);
+                    if (originalField == null) {
+                        return;
+                    }
+                    TypeMirror fieldType = determineFieldType(originalField, methodsPerAttribute.get(originalField));
+                    String field = fieldsClassName + "." + toEnumConstant(originalField);
```
Comment (vramik) [inline model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:549]: Are those indentation changes intentional?

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/Util.java [hunk 2]
```diff
@@ -46,7 +47,7 @@
 public class Util {
 
     private static final Set<String> LIST_TYPES = Set.of(List.class.getCanonicalName(), ArrayList.class.getCanonicalName(), LinkedList.class.getCanonicalName());
-    private static final Set<String> SET_TYPES = Set.of(Set.class.getCanonicalName(), TreeSet.class.getCanonicalName(), HashSet.class.getCanonicalName(), LinkedHashSet.class.getCanonicalName());
+    private static final Set<String> SET_TYPES = Set.of(Set.class.getCanonicalName(), TreeSet.class.getCanonicalName(), HashSet.class.getCanonicalName(), LinkedHashSet.class.getCanonicalName(), Collection.class.getCanonicalName());
```
Comment (vramik) [inline model/build-processor/src/main/java/org/keycloak/models/map/processor/Util.java:50]: Could you please elaborate why is the `Collection` considered as set type here?

File: model/map-file/pom.xml [hunk 3]
```diff
@@ -28,15 +28,29 @@
...
             <artifactId>keycloak-model-map</artifactId>
         </dependency>
+
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <scope>test</scope>
+            <groupId>org.snakeyaml</groupId>
+            <artifactId>snakeyaml-engine</artifactId>
+            <version>2.5</version>
```
Comment (vramik) [inline model/map-file/pom.xml:53]: Shouldn't we declare the version (and maybe the dependency as well) within `keycloak/pom/xml`?

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileCriteriaBuilder.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileCriteriaBuilder.java:2]: nitpick, it applies to other new files as well
```suggestion
 * Copyright 2023 Red Hat, Inc. and/or its affiliates
```

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+/**
+ * A class implementing a {@code YamlContext} interface represents a transformer
+ * from a primitive value / sequence / mapping representation as declared in YAML
+ * format into a Java object of type {@code V}, with ability to produce
+ * the {@link #getResult() resulting instance} of parsing.
+ *
+ * <p>
+ * NOTE: In the future, this transformer might also cover the other direction:
+ * conversion from Java object into YAML primitive value / sequence / mapping representation.
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java:37]: Isn't this note outdated? It seems to me that with `void writeValue(V value, WritingMechanism mech);` method it covers the other direction already.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import static org.keycloak.models.map.common.CastUtils.cast;
+
+/**
+ * A class implementing a {@code YamlContext} interface represents a transformer
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java:30]:
```suggestion
 * A class implementing a {@code BlockContext} interface represents a transformer
```

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+ * from a primitive value / sequence / mapping representation as declared in YAML
+ * format into a Java object of type {@code V}, with ability to produce
+ * the {@link #getResult() resulting instance} of parsing.
+ *
+ * <p>
+ * NOTE: In the future, this transformer might also cover the other direction:
+ * conversion from Java object into YAML primitive value / sequence / mapping representation.
+ *
+ * <p>
+ * This transformer handles only a single nesting level in YAML file. The root level
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java:40]: I don't fully understand this statement. Could you please elaborate a little bit? Maybe an example where the "single nesting level" could be seen, would help to understand. Thank you

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/StringListMapContext.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        return new Prefixed(prefix);
+    }
+
+    public static DefaultMapContext singletonAttributesMap(String key) {
+        return new SingletonAttributesMapYamlContext(key);
+    }
+
+    @Override
+    public AttributeValueYamlContext getContext(String nameOfSubcontext) {
+        // regardless of the key name, the values need to be converted into Set<String> which is the purpose of AttributeValueYamlContext
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/StringListMapContext.java:55]:
```suggestion
        // regardless of the key name, the values need to be converted into List<String> which is the purpose of AttributeValueYamlContext
```

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/WritingMechanism.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package org.keycloak.models.map.storage.file.common;
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/WritingMechanism.java:1]: missing copyright

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+        return (E) contextStack.pop().getResult();
+    }
+
+    protected Object parseNode() {
+        if (parser.checkEvent(Event.ID.Alias)) {
+            throw new IllegalStateException("Aliases are not handled at this moment");
+        }
+        Event ev = parser.next();
+//        System.out.println("  Parsing " + ev);
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java:128]: Do we want to keep the commented out lines in the file? There are multiple occurences.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlWritingMechanism.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
+package org.keycloak.models.map.storage.file.yaml;
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlWritingMechanism.java:1]: missing copyright

**Review by vramik — COMMENTED** [review]
Thank you @hmlnarik. It's great to see it. I like it very much.

I'd like to take a closer look at it, in the meantime I went through the code and I have few preliminary questions, please see inline.

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java (hunk 1 above)
Comment (hmlnarik) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:549]: nope, thanks for spotting it

File: model/map-file/pom.xml (hunk 3 above)
Comment (hmlnarik) [reply model/map-file/pom.xml:53]: Good catch

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java (hunk 7 above)
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/BlockContext.java:40]: I will update the javadoc in that regard, to specify what is meant by the "nesting level".

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java (hunk 10 above)
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java:128]: No, they will be removed

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public void commit() {
+        super.commit();
+        renameOnCommit.forEach(FileMapKeycloakTransaction::silentMove);
+        this.touchedPaths.forEach(FileMapKeycloakTransaction::silentDelete);
+    }
+
+    private static void silentMove(Path from, Path to) {
```
Comment (ahus1) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java:87]: Those `silent()` methods swallow exceptions, which can be dangerous. I don't see any JavaDoc indicating when it should be used and when not. Please advise when it is a good idea to swallow those. I just assume bad things can happend when something isn't deleted that is supposed to be deleted, or not moved when it was supposed to be moved.

File: model/map-file/pom.xml [hunk 13]
```diff
@@ -28,15 +28,29 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>2.3.2</version>
+                <configuration>
+                    <source>11</source>
+                    <target>11</target>
+                </configuration>
```
Comment (ahus1) [inline model/map-file/pom.xml:40]: I'm surprise to see the compiler plugin's version and source/target defined here. Shouldn't this be the same for all Keycloak? Could you please add a comment with maybe the error message you were seeing without this?

**Review by ahus1 — CHANGES_REQUESTED** [review]
Hi @hmlnarik - happy to see this PR to be now available!

See below some comment from reading the code. I'm usually a big fan of class-level-JavaDoc, so I asked for it in some places where there are new classes.

In some places my IDE showed me warnings around generics, that for example `BlockContext` could be turned into `BlockContext<?>`. I didn't mark all of those. Please let me know how you think about them.

File: model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java [hunk 14]
```diff
@@ -154,6 +178,53 @@ public static <M> Class<M> getModelClass(String key) {
...
+        return ENTITY_FIELDS.containsKey(entityClass);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <V> Stream<EntityField<V>> getEntityFields(Class<V> entityClass) {
+        EntityField<V>[] values = (EntityField<V>[]) ENTITY_FIELDS.get(entityClass);
+        return values == null ? Stream.empty() : Stream.of(values);
+    }
+
+    public static <V extends AbstractEntity> Optional<EntityField<V>> getEntityField(Class<V> entityClass, String fieldNameCamelCase) {
```
Comment (martin-kanis) [inline model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java:190]: Unused method. Is it something for the future or it can be removed?

File: model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java [hunk 15]
```diff
@@ -154,6 +178,53 @@ public static <M> Class<M> getModelClass(String key) {
...
+        final Stream<EntityField<V>> s = getEntityFields(entityClass);
+
+        return s
+          .filter(ef -> fieldNameCamelCase.equals(ef.getNameCamelCase()))
+          .findAny();
+    }
 
+    @SuppressWarnings("unchecked")
+    public static <V extends AbstractEntity> EntityField<V> getIdField(Class<V> targetEntityClass) {
```
Comment (martin-kanis) [inline model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java:199]: Unused method. Is it something for the future or it can be removed?

File: model/map-file/pom.xml (hunk 13 above)
Comment (hmlnarik) [reply model/map-file/pom.xml:40]: Good spot. This is partly correct - the compiler plugin remained here due to historical reasons and will be fixed with the new commit.
As for forcing the JDK version, it is set to be 11 (due to Keycloak source requirement inconsistency) similarly to e.g. quarkus module: https://github.com/keycloak/keycloak/blob/51bed81814574d254c07f091936483cf97716fa7/quarkus/pom.xml#L52-L54

File: model/map/src/main/java/org/keycloak/models/map/common/CastUtils.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        if (toClass == String.class) {
+            return Objects::toString;
+        }
+        if (fromClass == String.class) {
+            if (toClass == Integer.class) {
+                return (String s) -> Integer.valueOf(s);
+            } else if (toClass == Long.class) {
+                return (String s) -> Long.valueOf(s);
+            } else if (toClass == Boolean.class) {
+                return (String s) -> Boolean.valueOf(s);
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/common/CastUtils.java:46]: It's a bit trickier as it needs explicit casting, but I agree that method reference is better. Without `(Function<String, ?>)` cast, compiler raises the following error:
```
incompatible types: invalid method reference
    no suitable method found for valueOf(java.lang.Object)
        method java.lang.Boolean.valueOf(java.lang.String) is not applicable
          (argument mismatch; java.lang.Object cannot be converted to java.lang.String)
        method java.lang.Boolean.valueOf(boolean) is not applicable
          (argument mismatch; java.lang.Object cannot be converted to boolean)
1 error
```

File: model/map/src/main/java/org/keycloak/models/map/common/HasRealmId.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.keycloak.models.map.common;
+
+/**
+ *
+ * @author hmlnarik
+ */
+public interface HasRealmId {
+    
+    String getRealmId();
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/common/HasRealmId.java:25]: > So I wonder if this method can be removed.

The interface needs to retain both methods.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java (hunk 12 above)
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java:87]: These are private methods of `FileMapKeycloakTransaction`. I've updated the `silentMove` to throw an exception if it cannot finish its job as it can spoil the filesystem, but failed removal does not lead to inconsistency.

File: model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java (hunk 14 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java:190]: It's something for the future, to be used later in the tree store.

File: model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java (hunk 15 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/ModelEntityUtil.java:199]: It's something for the future, to be used later in the tree store.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapStorageProviderFactory.java [hunk 18]
```diff
@@ -17,13 +17,44 @@
 package org.keycloak.models.map.storage.file;
 
 import org.keycloak.Config;
+import org.keycloak.authorization.model.PermissionTicket;
+import org.keycloak.authorization.model.Policy;
+import org.keycloak.authorization.model.Resource;
+import org.keycloak.authorization.model.ResourceServer;
```
Comment (vramik) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapStorageProviderFactory.java:23]: unused imports

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you @hmlnarik, great job!

I have noticed that there are almost all areas (except singleUseObjects) available. 

I was wondering whether it's desired to have areas like auth-sessions or user-sessions in file store. 

The question is motivated by the other task: https://github.com/keycloak/keycloak/issues/16821... i.e. what configuration do we want to have for Quarkus?

Also I've noticed that there are not used any specific contexts, like [`ProtocolMappersContext`](https://github.com/keycloak/keycloak-playground/blob/6a7de121d58d9760523b8d80edb7c2efa5da1d86/file-store/src/main/java/org/keycloak/models/map/storage/file/client/ProtocolMappersYamlContext.java) nor [`Prefixed`](https://github.com/keycloak/keycloak/pull/16001/files#diff-6266e94fc7537e85b1a0c0247f16aa2ea276fe48079ee2aba85f67c35b2a9496R46) and [`SingletonAttributesMapYamlContext`](https://github.com/keycloak/keycloak/pull/16001/files#diff-6266e94fc7537e85b1a0c0247f16aa2ea276fe48079ee2aba85f67c35b2a9496R50). I assume it'll be handled within some other follow-up tasks, won't it?

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/MapEntityContext.java [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+            Class<?> collectionElementClass = ef.getCollectionElementClass();
+            ((Collection) value).forEach(v -> ef.collectionAdd(result, cast(v, collectionElementClass)));
+        } else if (ef.getMapKeyClass() != Void.class && value instanceof Map) {
+            Class<?> mapKeyClass = ef.getMapKeyClass();
+            Class<?> mapValueClass = ef.getMapValueClass();
+            ((Map) value).forEach((k, v) -> ef.mapPut(result, cast(k, mapKeyClass), cast(v, mapValueClass)));
+        } else {
+            final Object origValue = ef.get(result);
+            if (origValue != null) {
+                LOG.warnf("Overwriting value of %s field", ef.getNameCamelCase());
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/common/MapEntityContext.java:138]: Yes, this is expected. This does not block the functionality but it is annoying log message. Therefore will be a follow-up issue to fix it.

The original approach in this PR (change the `MapRoleEntity` as you suggested) lead to failures in HotRod which expects the current behaviour and the fix will not be completely straightforward.

**Comment by hmlnarik** [comment]
> I was wondering whether it's desired to have areas like auth-sessions or user-sessions in file store.

Yes, there is in this PR which aims to deliver generic basis for all entitites. In real-world usecases, session data will likely be covered by another technology capable of clustering.

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/Util.java (hunk 2 above)
Comment (hmlnarik) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/Util.java:50]: Is there a value of distinguishing between `Collection` and `Set`?
If not (and I don't see it here), I would prefer keeping it as is.

**Review by vramik — DISMISSED** [review]

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank for updating the first round of changes. In this review phase, I did a smoke test on the file store. What I found in addition to the code comments below: 

* When I setup a new Keycloak instance, I see the warning "Overwriting value of clientRole field" - I suppose I shouldn't see that warning. Also see it after logging in for the admin console
* When running it with the Maven command, it stores its files in the '/target/file' relative to the root folder. AFAIK the CHM would store them relative to the target folder of the current module, so I would have expected them below `testsuite/utils/target/file`
* I found some possible ambiguities when saving user attributes which are AFAIK strings: 
  * A value `1.000` entered in the frontend was stored in YAML as `1.000` (no quotes). When I then retrieved it for the frontend, it showed as `1.0` (trailing zeroes removed), as YAML now reads it back as a floating point number. IMHO similar things would apply to numbers.
  * A value `true` entered in the frontend was stored in YAML as `true` (no quotes). In YAML, this is no longer a string, but a boolean. Reading the values for the frontend again, is showed up without issues - for other tools which might expect a string and not a boolean value, this might come as a surprise. I would expect Keycloak to write that value with quotes to YAML. Unfortunately there are lots of values YAML treats as booleans when they are not quotes.
* I tried to put YAML references into the strings with `&` and `*` and Keycloak added quotes around them as expected - nothing that needs changing in that area.

As some extended test, created a user with the name `1.0000` which then showed up later as `1.0` in the UI. When trying to modify it, it mentioned that the user was not found. This applies to the pattern seen above.

Suggestion for the ambiguities: I'd expect Keycloak to add more often quotes around the values written to the store when the String would otherwise be interpreted as a number or boolean value when reading the YAML later. 

I created a user with the name `../test` which created a file outside of the users folder. The user was in the next step then not found. Apparently some validation/sanitation is missing on IDs. This could be used to overwrite arbitrary files on the server's file system.

I suggest a very restricted charset (`[-a-zA-Z0-9_]`), with also prohibiting UTF-8 characters here. I know this might limit some scenarios here (people having an umlaut in their user name), still I consider this both a security measure, and something to prevent obscure bugs. As those IDs are also used in URLs, at a minimum prohibit all  unsafe and reserved characters (found this entry on [SO](https://stackoverflow.com/questions/695438/what-are-the-safe-characters-for-making-urls), still this list might not be complete).
A different safe approach would be to use UUIDs for the IDs users, clients, roles etc. 

UPDATE: using email addresses as IDs requires at least `.` and `@`. Also some other characters like `+` are allowed in email addresses.

UPDATE2: maybe URL-encode the names before using them as file names (including escaping of all UTF-8 characters)?

UPDATE3: There are case sensitivity issues when running on Windows. This would relate to those file name issues described here.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+        }
+        Event ev = parser.next();
+        if (!(ev instanceof NodeEvent)) {
+            throw new IllegalArgumentException("Invalid event " + ev);
+        }
+//        if (anchor != null) {
+//            node.setAnchor(anchor);
+//            anchors.put(anchor, node);
+//        }
+//        try {
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java:135]: This is intended, there might be support for anchors added in the future and this part originates in snakeyaml

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    private static void silentDelete(Path p) {
+        try {
+            if (Files.exists(p) && Files.size(p) == 0) {
+                Files.delete(p);
+            }
+        } catch (IOException e) {
+            // Swallow the exception
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java:102]: This is _silent_ delete and the nonexistence of the file is expected, thus no exception should be thrown.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+            case SequenceStart:
+                return parseSequence();
+            case MappingStart:
+                return parseMapping();
+            default:
+                throw new IllegalStateException("Event not expected " + ev);
+        }
+//        } finally {
+//            anchors.remove(anchor);
+//        }
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/yaml/YamlParser.java:151]: This is intended, there might be support for anchors added in the future and this part originates in snakeyaml

**Comment by hmlnarik** [comment]
Thank you for the code review.

> Thank for updating the first round of changes. In this review phase, I did a smoke test on the file store. What I found in addition to the code comments below:
> 
> * When I setup a new Keycloak instance, I see the warning "Overwriting value of clientRole field" - I suppose I shouldn't see that warning. Also see it after logging in for the admin console

This is expected. Please see the description and #16857.

> * When running it with the Maven command, it stores its files in the '/target/file' relative to the root folder. AFAIK the CHM would store them relative to the target folder of the current module, so I would have expected them below `testsuite/utils/target/file`


> * I found some possible ambiguities when saving user attributes which are AFAIK strings:
>   
>   * A value `1.000` entered in the frontend was stored in YAML as `1.000` (no quotes). When I then retrieved it for the frontend, it showed as `1.0` (trailing zeroes removed), as YAML now reads it back as a floating point number. IMHO similar things would apply to numbers.

This needs fixing.

>   * A value `true` entered in the frontend was stored in YAML as `true` (no quotes). In YAML, this is no longer a string, but a boolean. Reading the values for the frontend again, is showed up without issues - for other tools which might expect a string and not a boolean value, this might come as a surprise. I would expect Keycloak to write that value with quotes to YAML. Unfortunately there are lots of values YAML treats as booleans when they are not quotes.

This is intended

> * I tried to put YAML references into the strings with `&` and `*` and Keycloak added quotes around them as expected - nothing that needs changing in that area.
> 
> As some extended test, created a user with the name `1.0000` which then showed up later as `1.0` in the UI. When trying to modify it, it mentioned that the user was not found. This applies to the pattern seen above.
> 
> Suggestion for the ambiguities: I'd expect Keycloak to add more often quotes around the values written to the store when the String would otherwise be interpreted as a number or boolean value when reading the YAML later.
> 
> I created a user with the name `../test` which created a file outside of the users folder. The user was in the next step then not found. Apparently some validation/sanitation is missing on IDs. This could be used to overwrite arbitrary files on the server's file system.

Please see description. Path traversal is known and will be treated in a follow-up issue.

> I suggest a very restricted charset (`[-a-zA-Z0-9_]`), with also prohibiting UTF-8 characters here. I know this might limit some scenarios here (people having an umlaut in their user name), still I consider this both a security measure, and something to prevent obscure bugs. As those IDs are also used in URLs, at a minimum prohibit all unsafe and reserved characters (found this entry on [SO](https://stackoverflow.com/questions/695438/what-are-the-safe-characters-for-making-urls), still this list might not be complete). A different safe approach would be to use UUIDs for the IDs users, clients, roles etc.

There will be URL-like escaping in place for the reserved characters of filesystems. UTF-8 characters will not be restricted since it generally is not restricted on filesystems and would make filenames much harder to read which violates the aim of the file store. There is [`TODO`](https://github.com/keycloak/keycloak/pull/16001/files#diff-9cc35d9e51a93c8a8cea73bc29368e7d3ac2121c665f821d6e12c43fb655a5dcR177-R183) in this effect.

> UPDATE: using email addresses as IDs requires at least `.` and `@`. Also some other characters like `+` are allowed in email addresses.
> 
> UPDATE2: maybe URL-encode the names before using them as file names (including escaping of all UTF-8 characters)?
> 
> UPDATE3: There are case sensitivity issues when running on Windows. This would relate to those file name issues described here.

It is not possible to create two files with names differing in lettercase only in Windows. Attempt to create such an object (say `admin` and `aDmin` user) would fail - just the same as it would in case of LDAP in AD.

**Comment by ahus1** [comment]
@hmlnarik 

> This is not-correct, CHM stores them relative to the root folder (or any folder it is run from)

Thanks for pointing me to the `keycloak-server.json`. I now see that neither the property `keycloak.group.map.storage.provider.directory` (used by file store) nor the property  `project.build.directory` (used by CHM) are set when running via maven. Should aligned outside of this PR.

> [handling of 1.000] - This needs fixing. EDIT: Could you be more specific where you put the value to?

I put this in two places: 

* as a generic attribute value for an existing user -> I saw the value without the zeros at the end when reloading the page
* as a user name for a new user -> I saw an error message when trying to update the user

This could be a separate issue to be followed-up on later.

>> A value true entered in the frontend was stored in YAML as true (no quotes).

> This is intended

Could you please explain scenarios when this would be intended? I still think this would be a very surprising behavior when editing such a YAML file or processing it with other tools, as those might treat `true` and `Y` as the same boolean value, while Keycloak seems to treat them differently. We could have this as a separate discussion after this PR being merged.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (ahus1) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java:102]: **TL;DR**: Doing some search where this is being called from and when, this seems mostly harmless. Still, this swallowed exception will never occur when the file has already been deleted. 

In a previous review I asked for some comments when then this method should be used. Tracing this back it seems only be used for `touchedPaths`. Looking where this is being filled, it seems to used for temporary files. Looking at the body of the method, it also handles only empty files. It also checks if the file exists. 

A deletion of a file might fail due to multiple reasons, not only because it no longer exists. The file system might be a network file system and the operation might temporarily fail. Other reasons would be that the file system might be mounted read-only, the Keycloak instance might lack the required permissions (unlikely here as Keycloak created the files earlier in the transaction).  In such a situation, when the file exists, but can't be deleted, throwing an exception is IMHO valid. Not sure when those temporary files would cause problems later on. 

So to summarize, looking at the code, there should be no `IOException` when the file doesn't exist as it is handled via `Files.exists(p)`. There might be other `IOException`s for other reasons as described above, although I don't know if those would be harmful here / would lead to problems eventually when the file is not removed. 

So it is still not obvious to me why it is either important to swallow the exception here. 

The case where a delete that isn't executed is problematic is then probably here in `FileMapStoreage.Crud#delete()` where there is currently only warning.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    public void touch(Path path) throws FileAlreadyExistsException, IOException {
+        Files.createFile(path);
+        touchedPaths.add(path);
+    }
+
+    void registerRenameOnCommit(Path from, Path to) {
+        this.renameOnCommit.put(from, to);
+        this.touchedPaths.add(from);
```
Comment (ahus1) [inline model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java:113]: When debugging this, I was surprised that the `from` path was added to the `touchedPaths`. 

As the `renameOnCommit` is executed first on the happy commit path, the removing of all touched paths doesn't do anything on the `fromfile` - it is already moved to the new name.

On the unhappy path on rollback, that `from` file will only be removed when it is still empty. It's a bit hard for me to follow on this logic to see if this leads always to the correct results, and if the deletion of files on rollback could be determined in a simpler way.

**Comment by hmlnarik** [comment]
> This should be unblocked by #12462. Should it also be created follow-up task for it?

No, there will be some additional work around realm-specific stores which would clash with this one.

File: model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java (hunk 23 above)
Comment (hmlnarik) [reply model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapKeycloakTransaction.java:113]: `from` files are temporary files owned by current transaction, e.g. `.tx-123-realm.yaml`. They have to be removed on commit (via renaming: `silentMove`) or on rollback (deleting).

**Comment by ahus1** [comment]
> _Sidenote:_ `Y` is not valid YAML boolean value, see https://yaml.org/spec/1.2.2/#10212-boolean which only allows `true` and `false`. Keycloak adheres to YAML spec as implemented by snakeyaml-engine.

Oh, what a great surprise to have snakeyaml implement YAML 1.1 and snakeyaml-engine implement YAML 1.2!

**Review by ahus1 — DISMISSED** [review]
Thank you for implementing this base for the file store. Looking forward to the things to come!

**Comment by hmlnarik** [comment]
Rebased

**Comment by martin-kanis** [comment]
When I start Keycloak (keycloak-server from testsuite) and log in as an admin, then log out, log in again and so on, I see that user session are not being removed. I see multiple user session yaml files and also multiple active sessions under "sessions" tab in admin console. Is this expected?
When trying with pure CHM (with additional server's restarts to persist data in files), I don't see multiple user session inside map-user-sessions.json

**Comment by hmlnarik** [comment]
> When I start Keycloak (keycloak-server from testsuite) and log in as an admin, then log out, log in again and so on, I see that user session are not being removed. I see multiple user session yaml files and also multiple active sessions under "sessions" tab in admin console. Is this expected? When trying with pure CHM (with additional server's restarts to persist data in files), I don't see multiple user session inside map-user-sessions.json

@martin-kanis Thank you, this has been fixed in the last commit (b47cbe8).

File: testsuite/utils/src/main/resources/META-INF/keycloak-server.json [hunk 24]
```diff
@@ -17,10 +17,10 @@
         },
         "map": {
             "storage-admin-events": {
-                "provider": "${keycloak.adminEventsStore.map.storage.provider:concurrenthashmap}"
+                "provider": "${keycloak.adminEventsStore.map.storage.provider,keycloak.mapStorage.provider.default:concurrenthashmap}"
```
Comment (martin-kanis) [inline testsuite/utils/src/main/resources/META-INF/keycloak-server.json:20]: Should this changes go into `testsuite/integration-arquillian/tests/base/src/test/resources/META-INF/keycloak-server.json` as well?

Comment (martin-kanis) [inline testsuite/utils/src/main/resources/META-INF/keycloak-server.json:20]: Should this changes go into `testsuite/integration-arquillian/tests/base/src/test/resources/META-INF/keycloak-server.json` as well?

**Review by martin-kanis — APPROVED** [review]
@hmlnarik Thank you for the huge amount of work!

File: testsuite/utils/src/main/resources/META-INF/keycloak-server.json (hunk 24 above)
Comment (hmlnarik) [reply testsuite/utils/src/main/resources/META-INF/keycloak-server.json:20]: In the arquillian testsuite, there are `map-storage-*` profiles which define all the necessary properties for map storage. Here it is only for convenience usage from CLI. So while I agree it would be nice, I consider this a follow-up task if the need arises.
